core/classes/Adapter.js

```diff
@@ -200,7 +217,79 @@ module.exports = class Adapter {
       }
     })
 
-    return trx ? query.transacting(trx) : query
+    let result = await (trx ? query.transacting(trx) : query)
+    for (let field of otmPopulate) {
+      await this._oneToModelPopulate(result, field)
+    }
+    for (let field of mtmPopulate) {
+      await this._modelToModelPopulate(result, field.linkerModel, field.initialModel, field.modelToLink)
+    }
+    return result
+  }
+
+  /**
+   * Populates the given models with 1:m relation
+   * @param {Object[]} initialModelArr The array of models to be populated
+   * @param {String} populatingModelIdentity The identity of related model
+   * @private
+   */
+  async _oneToModelPopulate(initialModelArr, populatingModelIdentity) {
+    let idsSet = new Set()
+    for (let model of initialModelArr) {
+      idsSet.add(model[populatingModelIdentity])
+    }
+    const idsArr = [...idsSet]
+
+    const populatingModelsArr = await this.knex(populatingModelIdentity.toLowerCase()).whereIn("id", idsArr)
+    for (let model of initialModelArr) {
+      for (let populatingModel of populatingModelsArr) {
+        if (model[populatingModelIdentity] === populatingModel.id) {
+          model[populatingModelIdentity] = populatingModel
+          break
+        }
+      }
+    }
+  }
+
+  /**
+   * Populates the given model with m:m relation
+   * @param {Array} initialModelArr The array of models to be populated
+   * @param {String} linkerModelIdentity The identity of linker model
+   * @param {String} initialModelIdentity The identity of initial model
+   * @param {String} modelToLinkIdentity The identity of model to be linked
+   * @private
+   */
+  async _modelToModelPopulate(initialModelArr, linkerModelIdentity, initialModelIdentity, modelToLinkIdentity) {
+    let initialModelIdsSet = new Set()
+    for (let initialModel of initialModelArr) {
+      initialModelIdsSet.add(initialModel.id)
+      initialModel[modelToLinkIdentity] = []
+    }
+    const initialModelIdsArr = [...initialModelIdsSet]
+
+    const linkerModelArr = await this
+      .knex(linkerModelIdentity.toLowerCase())
+      .whereIn(initialModelIdentity, initialModelIdsArr)
+    let modelToLinkIdsSet = new Set()
+    for (let linkerModel of linkerModelArr) {
+      modelToLinkIdsSet.add(linkerModel[modelToLinkIdentity])
+    }
+    const modelToLinkIdsArr = [...modelToLinkIdsSet]
+
+    const modelToLinkArr = await this.knex(modelToLinkIdentity.toLowerCase()).whereIn("id", modelToLinkIdsArr)
+    for (let linkerModel of linkerModelArr) {
+      for (let initialModel of initialModelArr) {
+        if (initialModel.id === linkerModel[initialModelIdentity]) {
+          for (let modelToLink of modelToLinkArr) {
```

Comment:
Three inserted cycles are not good at all. Think of creating two id-based indexes of models and linked models before the main cycle and you'll avoid this bottleneck.